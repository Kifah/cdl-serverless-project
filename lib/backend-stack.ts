import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import {LambdaIntegration, RestApi} from "aws-cdk-lib/aws-apigateway";
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Runtime} from "aws-cdk-lib/aws-lambda";
import * as lambda from "aws-cdk-lib/aws-lambda";


export class BackendStack extends cdk.Stack {

    public readonly backendTable: dynamodb.ITable;

    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        //add dynamoDB table and give it streaming and permissions needed to forward messages to pre-process lambda
        this.backendTable = new dynamodb.Table(this, 'CdkPlayBackendTable', {
            partitionKey: {name: 'id', type: dynamodb.AttributeType.STRING},
            billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            stream: dynamodb.StreamViewType.NEW_IMAGE
        });


        const backendCarsLambda = new NodejsFunction(this, 'CdkPlayBackendLambda', {
            runtime: lambda.Runtime.NODEJS_18_X,
            tracing: lambda.Tracing.ACTIVE,
            entry: 'resources/backend/backend.ts',
            handler: "handler",
            environment: {
                TABLE_NAME:this.backendTable.tableName
            }
        });

        const lambdaIntegration = new LambdaIntegration(backendCarsLambda);


        const backendRestApi = new RestApi(this, 'CdkPlayBackendApi');
        const spacesResource = backendRestApi.root.addResource('cars');
        spacesResource.addMethod('GET', lambdaIntegration);
        spacesResource.addMethod('POST', lambdaIntegration);

        this.backendTable.grantReadWriteData(backendCarsLambda);


    }
}