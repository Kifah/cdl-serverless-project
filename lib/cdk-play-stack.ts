import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';


export class CdkPlayStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);



        //add dynamoDB table and give it streaming and permissions needed to forward messages to pre-process lambda
        const table = new dynamodb.Table(this, 'CdkPlayTable', {
            partitionKey: {name: 'id', type: dynamodb.AttributeType.STRING},
            stream: dynamodb.StreamViewType.NEW_IMAGE,
        });

        const queue = new sqs.Queue(this, 'CdkPlayQueue', {
            visibilityTimeout: cdk.Duration.seconds(300)
        });

        // a pre-processing lambda receives messages from dynamoDB
        const preProcessHandler = new lambda.Function(this, "CdkPlayPreProcessHandler", {
            runtime: lambda.Runtime.NODEJS_18_X,
            code: lambda.Code.fromAsset("resources"),
            handler: "preprocess.handler",
            environment: {
                SQS_URL: queue.queueUrl,
            }

        });

        const finalHandler = new lambda.Function(this, "CdkPlayHandler", {
            runtime: lambda.Runtime.NODEJS_18_X,
            code: lambda.Code.fromAsset("resources"),
            handler: "final.handler",

        });



        const tableEventSource = new lambdaEventSources.DynamoEventSource(table, {
            startingPosition: lambda.StartingPosition.LATEST,
            filters: [lambda.FilterCriteria.filter({eventName: lambda.FilterRule.isEqual('INSERT')})],
        });
        preProcessHandler.addEventSource(tableEventSource);
        table.grantStream(preProcessHandler);
        queue.grantSendMessages(preProcessHandler);
        queue.grantSendMessages(finalHandler);
        const sqsEventSource = new lambdaEventSources.SqsEventSource(queue);
        finalHandler.addEventSource(sqsEventSource);

    }
}
