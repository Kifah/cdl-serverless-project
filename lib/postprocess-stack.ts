import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {NodejsFunction} from 'aws-cdk-lib/aws-lambda-nodejs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import {join} from 'path';

interface PostprocessStackProps extends cdk.StackProps {
    backendTable: dynamodb.ITable
}

export class PostprocessStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: PostprocessStackProps) {
        super(scope, id, props);


        const queue = new sqs.Queue(this, 'CdkPlayQueue', {
            visibilityTimeout: cdk.Duration.seconds(300)
        });

        // a pre-processing lambda receives messages from dynamoDB
        const preProcessHandler = new NodejsFunction(this, "CdkPlayPreProcessHandler", {
            runtime: lambda.Runtime.NODEJS_18_X,
            entry: 'resources/preprocess.ts',
            handler: "handler",
            tracing: lambda.Tracing.ACTIVE,
            environment: {
                SQS_URL: queue.queueUrl,
            }
        });

        const postProcessHandler = new NodejsFunction(this, "CdkPlayHandler", {
            runtime: lambda.Runtime.NODEJS_18_X,
            tracing: lambda.Tracing.ACTIVE,
            entry: 'resources/postprocess.ts',
            handler: "handler",

        });


        const tableEventSource = new lambdaEventSources.DynamoEventSource(props.backendTable, {
            startingPosition: lambda.StartingPosition.LATEST,
            filters: [lambda.FilterCriteria.filter({eventName: lambda.FilterRule.isEqual('INSERT')})],
        });
        preProcessHandler.addEventSource(tableEventSource);
        props.backendTable.grantStreamRead(preProcessHandler);
        queue.grantSendMessages(preProcessHandler);
        queue.grantConsumeMessages(postProcessHandler);
        const sqsEventSource = new lambdaEventSources.SqsEventSource(queue);
        postProcessHandler.addEventSource(sqsEventSource);

    }
}
