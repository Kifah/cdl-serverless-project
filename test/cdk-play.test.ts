import * as cdk from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import * as CdkPlay from '../lib/cdk-play-stack';


const app = new cdk.App();
const stack = new CdkPlay.CdkPlayStack(app, 'MyTestStack');
const template = Template.fromStack(stack);

test('Has DynamoDB', () => {

    template.hasResourceProperties('AWS::DynamoDB::Table', {
        BillingMode: "PAY_PER_REQUEST"
    });

});

test('Has SQS', () => {

    template.hasResourceProperties('AWS::SQS::Queue', {
        VisibilityTimeout: 300
    });

});

test('Has Lambda with correct functions', () => {

    template.hasResourceProperties("AWS::Lambda::Function", {
        Handler: "preprocess.handler",
        Runtime: "nodejs18.x",
    });
    template.hasResourceProperties("AWS::Lambda::Function", {
        Handler: "final.handler",
        Runtime: "nodejs18.x",
    });

    template.resourceCountIs("AWS::Lambda::Function", 2);
});

test("Matches the snapshot", () => {
    expect(template.toJSON()).toMatchSnapshot();
});



