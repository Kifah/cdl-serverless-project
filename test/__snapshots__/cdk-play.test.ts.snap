// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CdkPlayHandler7964437C": {
      "DependsOn": [
        "CdkPlayHandlerServiceRoleDefaultPolicy37A2C2CC",
        "CdkPlayHandlerServiceRoleFD3E9DC6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5cc35fb487bb6e4a98d5ff78afc526f257a800ce10b2556eb2cc308900fbf9ba.zip",
        },
        "Handler": "final.handler",
        "Role": {
          "Fn::GetAtt": [
            "CdkPlayHandlerServiceRoleFD3E9DC6",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CdkPlayHandlerServiceRoleDefaultPolicy37A2C2CC": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkPlayQueue78BDD396",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkPlayHandlerServiceRoleDefaultPolicy37A2C2CC",
        "Roles": [
          {
            "Ref": "CdkPlayHandlerServiceRoleFD3E9DC6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkPlayHandlerServiceRoleFD3E9DC6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkPlayHandlerSqsEventSourceMyTestStackCdkPlayQueue308FB81D4203D444": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "CdkPlayQueue78BDD396",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "CdkPlayHandler7964437C",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "CdkPlayPreProcessHandler6CA76BA8": {
      "DependsOn": [
        "CdkPlayPreProcessHandlerServiceRoleDefaultPolicy8DACD1C1",
        "CdkPlayPreProcessHandlerServiceRole55175F02",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5cc35fb487bb6e4a98d5ff78afc526f257a800ce10b2556eb2cc308900fbf9ba.zip",
        },
        "Environment": {
          "Variables": {
            "SQS_URL": {
              "Ref": "CdkPlayQueue78BDD396",
            },
          },
        },
        "Handler": "preprocess.handler",
        "Role": {
          "Fn::GetAtt": [
            "CdkPlayPreProcessHandlerServiceRole55175F02",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CdkPlayPreProcessHandlerDynamoDBEventSourceMyTestStackCdkPlayTable61F3130FCB27F3E2": {
      "Properties": {
        "BatchSize": 100,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "CdkPlayTableF7283C8C",
            "StreamArn",
          ],
        },
        "FilterCriteria": {
          "Filters": [
            {
              "Pattern": "{"eventName":["INSERT"]}",
            },
          ],
        },
        "FunctionName": {
          "Ref": "CdkPlayPreProcessHandler6CA76BA8",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "CdkPlayPreProcessHandlerServiceRole55175F02": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkPlayPreProcessHandlerServiceRoleDefaultPolicy8DACD1C1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkPlayTableF7283C8C",
                  "StreamArn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkPlayQueue78BDD396",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkPlayPreProcessHandlerServiceRoleDefaultPolicy8DACD1C1",
        "Roles": [
          {
            "Ref": "CdkPlayPreProcessHandlerServiceRole55175F02",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkPlayQueue78BDD396": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkPlayTableF7283C8C": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`matches the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CdkPlayHandler7964437C": {
      "DependsOn": [
        "CdkPlayHandlerServiceRoleDefaultPolicy37A2C2CC",
        "CdkPlayHandlerServiceRoleFD3E9DC6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5cc35fb487bb6e4a98d5ff78afc526f257a800ce10b2556eb2cc308900fbf9ba.zip",
        },
        "Handler": "final.handler",
        "Role": {
          "Fn::GetAtt": [
            "CdkPlayHandlerServiceRoleFD3E9DC6",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CdkPlayHandlerServiceRoleDefaultPolicy37A2C2CC": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkPlayQueue78BDD396",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkPlayHandlerServiceRoleDefaultPolicy37A2C2CC",
        "Roles": [
          {
            "Ref": "CdkPlayHandlerServiceRoleFD3E9DC6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkPlayHandlerServiceRoleFD3E9DC6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkPlayHandlerSqsEventSourceMyTestStackCdkPlayQueue308FB81D4203D444": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "CdkPlayQueue78BDD396",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "CdkPlayHandler7964437C",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "CdkPlayPreProcessHandler6CA76BA8": {
      "DependsOn": [
        "CdkPlayPreProcessHandlerServiceRoleDefaultPolicy8DACD1C1",
        "CdkPlayPreProcessHandlerServiceRole55175F02",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5cc35fb487bb6e4a98d5ff78afc526f257a800ce10b2556eb2cc308900fbf9ba.zip",
        },
        "Environment": {
          "Variables": {
            "SQS_URL": {
              "Ref": "CdkPlayQueue78BDD396",
            },
          },
        },
        "Handler": "preprocess.handler",
        "Role": {
          "Fn::GetAtt": [
            "CdkPlayPreProcessHandlerServiceRole55175F02",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CdkPlayPreProcessHandlerDynamoDBEventSourceMyTestStackCdkPlayTable61F3130FCB27F3E2": {
      "Properties": {
        "BatchSize": 100,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "CdkPlayTableF7283C8C",
            "StreamArn",
          ],
        },
        "FilterCriteria": {
          "Filters": [
            {
              "Pattern": "{"eventName":["INSERT"]}",
            },
          ],
        },
        "FunctionName": {
          "Ref": "CdkPlayPreProcessHandler6CA76BA8",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "CdkPlayPreProcessHandlerServiceRole55175F02": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkPlayPreProcessHandlerServiceRoleDefaultPolicy8DACD1C1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkPlayTableF7283C8C",
                  "StreamArn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkPlayQueue78BDD396",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkPlayPreProcessHandlerServiceRoleDefaultPolicy8DACD1C1",
        "Roles": [
          {
            "Ref": "CdkPlayPreProcessHandlerServiceRole55175F02",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkPlayQueue78BDD396": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkPlayTableF7283C8C": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
